src/Strategy/NuGet/ProjectAssetsJson.hs

```diff
@@ -68,46 +82,94 @@ analyze' file = do
 
 instance FromJSON DependencyInfo where
   parseJSON = withObject "Dependency" $ \obj ->
     DependencyInfo <$> obj .: "type"
       <*> obj .:? "dependencies" .!= Map.empty
 
-data NuGetDep = NuGetDep
-  { depName :: Text
-  , depVersion :: Text
-  , completeDepType :: Text
-  , completeDeepDeps :: Map.Map Text Text
-  }
-  deriving (Show)
+instance FromJSON ProjectAssetsJson where
+  parseJSON = withObject "ProjectAssetsJson" $ \obj -> do
+    targets <- obj .: "targets"
+    projectFrameworks <- obj .: "project" |> "frameworks"
+    deps <- parseFramework projectFrameworks
+    pure $ ProjectAssetsJson targets deps
+    where
+      (|>) :: FromJSON a => Parser Object -> Text -> Parser a
+      (|>) parser key = do
+        obj <- parser
+        obj .: key
+
+      parseFrameworkDeps :: Value -> Parser (Set Text)
+      parseFrameworkDeps = withObject "parseFrameworkDeps" $ \o -> do
+        depsObj :: Maybe Object <- o .:? "dependencies"
+        deps <- case depsObj of
+          Nothing -> pure []
+          Just hm -> pure (fst <$> HM.toList hm)
+        pure $ Set.fromList deps
+
+      parseFramework :: Value -> Parser (Map.Map FrameworkName (Set Text))
+      parseFramework = withObject "parseFramework" $ \o -> do
+        projectFrameworks <- for (HM.toList o) $ \(framework, kv) -> do
+          frameworkDeps <- parseFrameworkDeps kv
+          pure (FrameworkName framework, frameworkDeps)
+        pure $ Map.fromList projectFrameworks
 
 buildGraph :: ProjectAssetsJson -> Graphing Dependency
-buildGraph project = unfold direct deepList toDependency
+buildGraph project = foldr (<>) Graphing.empty graphsOfTargetFrameworks
+  where
+    graphsOfTargetFrameworks =
+      map
+        (graphOfFramework $ projectFramework project)
+        (Map.toList $ targets project)
+
+graphOfFramework :: Map.Map FrameworkName (Set Text) -> (FrameworkName, Map.Map Text DependencyInfo) -> Graphing Dependency
+graphOfFramework projectFrameworkDeps (targetFramework, targetFrameworkDeps) =
+  Graphing.gmap toDependency
+    . withoutProjects
+    . promoteToDirect isDirectDep
+    $ unfoldDeep allResolvedDeps getTransitiveDeps id
   where
-    direct :: [NuGetDep]
-    direct = concatMap (mapMaybe convertDep . Map.toList) (Map.elems (targets project))
+    withoutProjects :: Graphing NuGetDep -> Graphing NuGetDep
+    withoutProjects = Graphing.shrink (not . isProjectDep)
+
+    isDirectDep :: NuGetDep -> Bool
+    isDirectDep d = depName d `elem` (getProjectDirectDepsByFramework)
 
-    convertDep :: (Text, DependencyInfo) -> Maybe NuGetDep
-    convertDep (depString, dep) = case Text.splitOn "/" depString of
+    isProjectDep :: NuGetDep -> Bool
+    isProjectDep NuGetDep{..} = completeDepType == "project"
+
+    allResolvedDeps :: [NuGetDep]
+    allResolvedDeps = mapMaybe toNugetDep $ Map.toList targetFrameworkDeps
+
+    toNugetDep :: (Text, DependencyInfo) -> Maybe NuGetDep
+    toNugetDep (depString, dep) = case Text.splitOn "/" depString of
```

Comment:
comment I forgot to add yesterday: we can do this check at JSON-decode-time, which would prevent us from silently ignoring invalid dependency keys -- and would also correctly propagate the error

there's a typeclass [called `FromJSONKey`](https://hackage.haskell.org/package/aeson-2.0.2.0/docs/Data-Aeson.html#t:FromJSONKey) that's used by `Map`'s `FromJSON` instance:

```hs
instance (FromJSONKey k, Ord k, FromJSON v) => FromJSON (Map k v) where
  ...
```

we can take advantage of this:

```hs
data NuGetDepKey = NuGetDepKey
  { depKeyName :: Text
  , depKeyVersion :: Text
  } deriving (Eq, Ord, Show)

instance FromJSONKey NuGetDepKey where
  fromJSONKey = FromJSONKeyTextParser $ \depString ->
    case Text.splitOn "/" depString of
      [name, ver] -> pure $ NuGetDepKey name ver
      other -> fail $ "Unexpected dependency key: " <> show depString
```

and change the type of `targets` in `ProjectAssetsJson`:

```hs
data ProjectAssetsJson = ProjectAssetsJson
  { targets :: Map.Map FrameworkName (Map.Map NuGetDepKey DependencyInfo)
  ...
```